source/transfer_list.rst

```diff
@@ -66,36 +66,31 @@ version field determines the contents of the handoff start header.
      - 0x4
      - The checksum is used to ensure the data contained within the list is intact. The checksum is set to a value such that the sum over every byte in the {tl_base_pa, …, tl_base_pa + size -1} address range, modulo 256, is equal to 0.
 
-   * - reserved
-     - 0x3
-     - 0x5
-     - Reserved, must be zero.
-
    * - version
-     - 0x4
-     - 0x8
+     - 0x1
```

Comment:
I think this commit is OK. There is quite a bit of unused space in the TL header and we have not come up with anything yet that should use it.

The complexity increase is minimal, as it needs to be, since saving 16 bytes is barely worth worrying about

Reply:
Or you could just swap the hdr_size and reserved fields so the version can be naturally extended later.

Reply:
I just realised in actually implementing something like this that we need a TE hdr_size of more than one byte. The header size is how you set the position of the data in the TE, so having a single byte limits the alignment to 128 bytes. Some things will want alignment of 64 bytes (ACPI table) or even 4KB (caches / page tables).

We could use a log (base 2) value but of course that involves more code[1] which defeats the purpose of the smaller TE header. It also doesn't solve the alignment issue since we may need to have a header of (e.g.) 40 bytes to make the data align on a 32-byte boundary, depending on what comes before.

So I believe we should stick with the 16-byte TE header. Going with 8 bytes provides no future space for flags or anything else, meaning we will have to update the version (a costly process in terms of code size) or force users to put extra bits in the TE data itself, even if they turn out to be common.

I also don't think saving this little space is worth it. From what I understand, TF-A has very little data passed (a few records), so it doesn't matter. Even if coreboot uses this for sysinfo (instead of device tree), then it is not too bad. I see coreboot passing about 26 records of about 800 bytes total, so the larger header would increase this to 800 + 26 * 8 =~ 1KB. I believe the flexibility it worth it.

[1]
05: log
       arm: (for 1/643 boards) all +128.0 text +128.0
            snow           : all +128 text +128
               u-boot: add: 0/0, grow: 9/0 bytes: 124/0 (124)
                 function                                   old     new   delta
                 bloblist_addrec                            104     176     +72
                 bloblist_calc_chksum                        70      86     +16
                 bloblist_show_list                          96     104      +8
                 bloblist_resize                            164     172      +8
                 bloblist_next_blob                          28      32      +4
                 bloblist_find                               28      32      +4
                 bloblist_ensure_size_ret                    48      52      +4
                 bloblist_ensure                             30      34      +4
                 bloblist_add                                26      30      +4
(no errors to report)

Reply:
The thread here was originally about the TL header, not the TE header. I presume there are no issues with 1-byte for the TL hdr_size (i.e. no alignment requirements on the first TE's base address)?

Regarding the TE hdr_size, where is the 64 byte ACPI table requirement coming from? The ACPI TE already has some reserved data in it to make it 16 byte aligned. What's the problem with that and doing similar for other TE types with alignment requirements?

If we decide we don't want to put in reserved data for alignment, then your "log (base 2)" idea sounds OK to me. That larger alignment would only apply to the TEs that need it, so I don't see the code size concerns you have. I guess in that case, a spec version would define a minimum hdr_size (for writers) rather than a fixed hdr_size.

So I'm not seeing the need for going back to 16-byte TE headers yet.

Reply:
Regarding versioning, I've just realized we haven't said what constitutes a version increase. Obviously a change in the structures will require this, but what about just adding TE types? And does a compliant implementation have to be able to write/parse all TE types? That seems too much if there are loads of TE types that only specific vendor/firmware implementations care about. I think adding TE types should not increase the version in the TL header (though it may result in a new spec release), and it should be implementation defined which TE types are implemented. There may be a case for some TE types being mandatory (e.g. XFERLIST_VOID).

Reply:
1. If we want a certain alignment for the data in any TE, I think we should require that the TL has at least that alignment.
2. ACPI needs 64-byte alignment for the base tables, or at least some OSes do.
3. We can't really have the users doing their own alignment. It would be like implementing memalign() with malloc() except that you don't even get an aligned pointer from the API. It needs to be handled by the API, so needs to be included in the TE header, from what I can see
4. The log2 thing doesn't help you if you are trying to align the data to 0x40 bytes. Say the TE header is at 0x10, the data at 0x20 (with the current 16-byte header). You need the data to start at 0x40, so the TE header size is 0x30, not a power of two.
5. Anyway my bigger point is that interoperability and flexibility is key for something like this. I gave some examples as to why I don't think we should fixate on the TE header size.
6. It should be OK to ignore types that you don't understand. I see the version number in the TL header being about the overall structure, not its contents.

Reply:
> If we want a certain alignment for the data in any TE, I think we should require that the TL has at least that alignment.

I initially thought you meant alignment of the TL hdr_size (i.e. the first TE's base addr) but I now see you mean alignment of the TL base address. I guess if we mandate this, it would avoid data getting misaligned when copying TLs around. But what if a TE with a higher alignment requirement is added to a pre-existing TL? The TL base address would suddenly be misaligned and it would be difficult to copy the TL somewhere else and maintain data alignment. I'm now wondering whether we need to store the maximum alignment requirement of all TEs in the TL header, and mandate that implementations maintain the TL base address offset from that max alignment?

The alternative would be to re-align all the data when copying TLs, but that would prevent bitwise copying of TLs and mean excluding the padding from the checksum calculation, which sounds nasty.

> ACPI needs 64-byte alignment for the base tables, or at least some OSes do.

OK, I'll take your word for that.

> We can't really have the users doing their own alignment. It would be like implementing memalign() with malloc() except that you don't even get an aligned pointer from the API. It needs to be handled by the API, so needs to be included in the TE header, from what I can see

I'm not sure I quite get the analogy but I think you're saying that if we do need to impose a TL base address alignment requirement (and maybe store a max alignment in the TL header), then it will be easier for common code to do the calculation if the TE alignment is stored in the common TE header instead of the specific TE. Assuming that it's not only the ACPI TE that has alignment requirements, then I agree with you.

> The log2 thing doesn't help you if you are trying to align the data to 0x40 bytes. Say the TE header is at 0x10, the data at 0x20 (with the current 16-byte header). You need the data to start at 0x40, so the TE header size is 0x30, not a power of two.

OK, I see your point now, but it would be possible to store both the (minimum) header size and the alignment in separate fields, with the latter being a power of 2. Actually, to cater for the problem you raise, I think this is needed anyway, since it would not be possible to deduce the alignment from the actual header size alone, e.g. the data may happen to be 1KB aligned but only require 64 byte alignment.

> Anyway my bigger point is that interoperability and flexibility is key for something like this. I gave some examples as to why I don't think we should fixate on the TE header size.

I agree interoperability and flexibility are key, but it's reasonable to try to keep the TE header size small if possible. You've convinced me that we need bigger headers, though.

> It should be OK to ignore types that you don't understand. I see the version number in the TL header being about the overall structure, not its contents.

+1. In that case, I'm less worried about blowing the 256 version limit, although if we end up with a bigger TL header, that issue goes away.

Let's see what others say after the holiday period.

Reply:
1/ To keep it simple, I think that alignment depends on the TL having the same or greater alignment. When relocating the TL, the alignment must not get 'worse'.
2. Yes that's what I am saying re having space in the common header.
3. Yes, the TE alignment needs to be specified by the TE header size, or at least that seems to be easiest.

Re holiday, yes, Merry Christmas!

Reply:
On TL versions:

Yes, of course the TL version shouldn't increase with every new tag, I thought that was a given. ;) I can add clarifying language for that in the next update. And of course implementations are not required to interpret any particular tag and should simply ignore tags they don't know. (I don't see any tag that would need to be mandatory -- e.g. the point of the VOID tag is just to ignore that area, so it is already treated the same as an unknown tag by readers.)

With that said, I assume 256 versions should be more than enough? I can imagine very few reasons why we'd ever want to increase the version anyway. The only reason to do that would be to add new fields to either the TL or TE headers -- there's only so much you can do with the TL header anyway, and the TE header is already full, so any new fields there would require to increase its size, and I'd have the same strong objections against that. I guess we could add some optional fields that only a few tags would use (and the remaining TE headers could stay 8 bytes long) or something? But I doubt that we'd go through more than 256 iterations of that.

On TE header size:

I think we have a big misunderstanding/desync in what the TE header size field should be used for here. In my understanding, the TE header size field only marks the area used by the generic, tag-independent TE header itself -- not for any tag-specific padding. For the current specification, the header size field _must be set to 8_ for all tags (that's how this PR writes it into all hdr_size field descriptions now). The generic TL structure itself controls the TE header, and the tag controls the insides of the TE data, and those two should not mix. The only thing that would allow different TE header sizes is if we updated the overall TL version and defined some new, tag-independent TE header fields.

Anything tag-specific, including alignment padding, needs to be sorted out within the TE data portion (that's covered by data_size, not hdr_size). That could be done either by starting the data portion with an explicit "offset until the real data starts" field, or just implicitly by alignment (e.g. the way I defined the ACPI table tag now is that the actual table data starts at the next 16-byte aligned address after the TE header). In that case the padding space between the end of the TE header and the aligned start of the actual data content counts as part of the "data" from the generic TL's point of view and is counted into data_size.

If we need to be able to support data aligned by more than 16 bytes, then I agree we should probably use the last reserved byte in the TL header to add a "minimum alignment (stored as power of 2) you need to place this TL at if you move it around" field. I can add that in the next update if we agree that we want this. (If you're trying to later add a TE with a large alignment requirement to a TL that was already placed at a smaller alignment, then... well... then you're out of luck and need to memmove() the whole thing first. I don't think we can have a perfect solution for every edge case. We could write something like "when creating or relocating a TL, the TL header should be placed at the largest aligned address that is convenient", but I don't think there should be any hard requirements beyond 16 byte.)

Reply:
For versions, yes I suspect 256 versions is enough and I agree it is hard to imagine why we would have to increase it. Although it always is at the beginning...

For TE header size, the current doc supports alignment as part of the structure. I don't want to lose that! It would be extremely odd IMO for users of the API to have to do something like:

void *blob = bloblist_find(TAG...);

// align to 64-bytes because that is what we need
blob = (void *)ALIGN((uintptr_t)blob, 64)

Also note that it is the TE data that is aligned, not the TE header (which would be pointless)

Re TL alignment, it might be useful, since otherwise it won't be obviously what alignment to use when relocating. In that case, a power of 2 is enough.

One option to support alignment and still have 8 byte TE headers would be to put a power of two in the TE header to indicate the alignment rather than the header size, with a minimum size of 8 bytes.

But why do we care? As I mentioned, it isn't a big deal. For coreboot we *really* don't care, since it is adding 200 bytes to an 800-byte structure in a system with GBs of memory, about to call its payload. I think it is better to play it safe and allow room.

Everything we do to make it more complicated to save 8 bytes per TE seems doomed to make us wish we hadn't in a few years.

So what is so important about having an 8-byte header?

Reply:
> For TE header size, the current doc supports alignment as part of the structure.

No, I don't think it does at all tbh, I find that a really weird misuse of the hdr_size field and if I read the spec even in its current form I would never assume it was legal to do that. Even the 16 byte header definition says hdr_size is "The size of this entry header in bytes.", and "The TE header is defined in :numref:`tab_te_header`.", where that table shows exactly 4 fields of 4 bytes each and absolutely no indication that you're just allowed to arbitrarily add extra padding fields below which the spec doesn't show.

A normal kind of API I would expect for this is
```
struct mytag *s = tl_find(TAG_MYTAG);
if (s)
  do_something_with(s->myfield);
```
Of course you can't do this in the ACPI case because an ACPI table is a weird, complex data structure with excessive alignment requirements that requires a complex parser to read, which is exactly why we are explicitly saying that new TE layouts should not work like that and should instead be small, simple, and representable by C structures. Since we do want to allow interoperability with these complex legacy structures, we have a way to make them work, but yeah then maybe you'll need to call one extra function to adjust the pointer in that case because excessive alignment requirements are a bad edge case and not the expected default for the majority of TEs. (You'll need some more complex handling for the ACPI tag anyway since it allows packing more than one table in the same TE with another alignment boundary in between, so you might as well build the initial padding adjustment into that code too.)

I find the suggestion to make every single TE header twice as bulky just so that we need to write one less line of code to access the one tag which represents an unusual interop-glue binding to a different kind of platform-specific handoff format with extreme placement requirements quite ridiculous.

Reply:
Re your first paragraph, we can easily add that information, but qually there is no indication that the only legal value for TE-header size if 16 bytes.

The API you describe would return a pointer to the data, not the TE header, right? The TE header is internal to the data structure so we don't want to expose that...it would make it hard to change!

It works fine for ACPI, so long as alignment is supported. You need 16- or perhaps 64-byte alignment, but that is it.

What is 'excessive' alignment?

Don't you think it is risky to use a header with absolutely no space for adjustments / expansion? I am still struggling to understand why saving 8 bytes is so important!

Reply:
> Re your first paragraph, we can easily add that information, but qually there is no indication that the only legal value for TE-header size if 16 bytes.

No, but I'd read it to say pretty clearly that the hdr_size just covers the header itself (which is made up of the listed fields), so I wouldn't expect that you can just arbitrarily make it bigger than that. And that's also the design that would make most sense to me. (Also, treating it this way would be required if we ever wanted to expand the TE header format with optional fields, which I think is probably the only good way to expand it at all if we ever see that need.)

> The API you describe would return a pointer to the data, not the TE header, right?

Yes, I mean that is up to every implementer and not defined by this specification, but I think a reasonable API would return the start of the data portion (since the header size cannot be assumed as compile-time constant, so you can't cover both header and data in a single struct).

> What is 'excessive' alignment?

Honestly, I think any beyond-natural alignment requirement is excessive. Does anyone know _why_ these ACPI tables need to be 64-byte aligned? I mean, they're not directly accessed by hardware, right? So is there any good reason for that or is that just something someone thoughtlessly made up 30+ years ago because they felt aligned things are cleaner or something?

But there's no real point arguing about that here, it is what it is and I recognize the need to make the ACPI table use case work somehow, of course. I just meant that this is an unusual edge case as opposed to a common case. We are unlikely to add many (any?) further TE types that require alignments that big. So I think it makes sense to hide the complications required due to this inside the ACPI tag rather than making special provisions for them in the generic part of the spec.

> Don't you think it is risky to use a header with absolutely no space for adjustments / expansion? 

I don't understand what you mean by "no space", we have a hdr_size field so we have up to 248 bytes for potential expansion? I'm just saying that making things bigger always comes with drawbacks so we should never do it lightly, and always try to look for options that allow adding the additional functionality for those who need it without burdening those who don't. In this case I think the solution I proposed makes the ACPI case work just as well without making things bigger for everyone else, so isn't that a clearly superior option?

Reply:
> > Re your first paragraph, we can easily add that information, but qually there is no indication that the only legal value for TE-header size if 16 bytes.
> 
> No, but I'd read it to say pretty clearly that the hdr_size just covers the header itself (which is made up of the listed fields), so I wouldn't expect that you can just arbitrarily make it bigger than that. And that's also the design that would make most sense to me. (Also, treating it this way would be required if we ever wanted to expand the TE header format with optional fields, which I think is probably the only good way to expand it at all if we ever see that need.)

OK, fair enough. It wasn't stated.. So we could add something like "the TE header size can be increased above the minimum of xx to achieve a particular alignment for TE data"

> 
> > The API you describe would return a pointer to the data, not the TE header, right?
> 
> Yes, I mean that is up to every implementer and not defined by this specification, but I think a reasonable API would return the start of the data portion (since the header size cannot be assumed as compile-time constant, so you can't cover both header and data in a single struct).

OK good

> 
> > What is 'excessive' alignment?
> 
> Honestly, I think any beyond-natural alignment requirement is excessive. Does anyone know _why_ these ACPI tables need to be 64-byte aligned? I mean, they're not directly accessed by hardware, right? So is there any good reason for that or is that just something someone thoughtlessly made up 30+ years ago because they felt aligned things are cleaner or something?

I think that 30 years ago a lot of things seemed like good ideas. Here we are presumably making some mistakes that we can only hope will last 30 years :-)

> 
> But there's no real point arguing about that here, it is what it is and I recognize the need to make the ACPI table use case work somehow, of course. I just meant that this is an unusual edge case as opposed to a common case. We are unlikely to add many (any?) further TE types that require alignments that big. So I think it makes sense to hide the complications required due to this inside the ACPI tag rather than making special provisions for them in the generic part of the spec.

Some examples I am aware of:
- ACPI is 64 bytes but typically we align it on a 4KB boundary
- SMBIOS is typically 256-byte boundary
- Chromium OS vboot context is 16 bytes
- Some things may want to be aligned on a cache boundary (e.g. 4KB), to enable locking in cache, or even holding a shared cache setup, etc. 

Given the importance of alignment, we could add a power-of-two field in the TE and TL headers to make it explicit?

> 
> > Don't you think it is risky to use a header with absolutely no space for adjustments / expansion?
> 
> I don't understand what you mean by "no space", we have a hdr_size field so we have up to 248 bytes for potential expansion? I'm just saying that making things bigger always comes with drawbacks so we should never do it lightly, and always try to look for options that allow adding the additional functionality for those who need it without burdening those who don't. In this case I think the solution I proposed makes the ACPI case work just as well without making things bigger for everyone else, so isn't that a clearly superior option?

It would be simpler to give ourselves a bit of space now, once we sort out the alignment question. Otherwise all the code needs to check for the two different types of TE header struct. What a pain!

I agree there may be drawbacks to using 8 bytes instead of 16, even if I am not sure what they might be. But I also feel that we should make this future-proof to at least some extent, so packing things tightly into 8 bytes seems like a premature optimisation.

Reply:
If we agree on the fundamental requirement to support multiple alignment sizes, then I propose the following (trying to accommodate everyone's needs):

- The TL version shouldn't increase with every new tag - we all seem agreed on that at least!
- Add a power-of-2 alignment field to the TL header (keeping it to 16 bytes). As I said before, I think we should mandate that implementations maintain the TL base address offset from that max alignment. This way, we don't need any particular alignment for the TL/TE base addresses. For larger alignments, I don't think it's reasonable to expect that the TL/TE base addresses have that alignment; we'd end up with huge amounts of padding after the TL header and between TEs. Also, this method allows you to add TEs with larger alignment requirements without requiring a memmove.
- Given that the ACPI TE is currently the only TE type with alignment requirements, don't add an alignment field to the TE header. Implementations that add ACPI TEs are expected to the set the TL header alignment field to 16 bytes.
- Don't use the TL hdr_size field for additional padding. I agree with Julius this would be unexpected usage and would conflict with any future expansion of the TL header with new fields.

I agree with Simon that there's a risk that some implementations may have to support another TE header struct sooner rather than later, but the spec caters for that and it doesn't sound that painful to me. Also, I don't see this as premature optimization - it's the only opportunity for optimization. We can't shrink things later.

Reply:
> As I said before, I think we should mandate that implementations maintain the TL base address offset from that max alignment.

Oh, sorry, I missed the "maintain the offset from that alignment when copying" part of your suggestion earlier. That's a great idea, then the alignment requirement can even be increased at a later point without requiring to move the whole list first. I'll incorporate that in the next update.

I think if we agree on that mechanism to handle any larger-than-natural alignments, it makes sense to reevaluate whether we even need to stick to the 16 byte TE alignment, or if we want to reduce the general alignment for all TE headers to 8. We originally argued for 16 just for the ACPI TE case, but since there are other use cases for even larger requirements now I don't think there's any reason to make special provisions for 16-byte ACPI alignments, they could just deal with it the same way as TEs with larger requirements. Lowering the general alignment to 8 would further save space for smaller elements.

As for how to align those TEs with larger requirements, a really simple solution that I didn't think of before would be to just not have any internal padding or extra "offset until the real data starts" fields at all, neither in the header nor the data portion of the TE. Instead, we could just mandate that whenever you write such a TE you have to ensure that the data portion starts at the correct alignment. If placing the new TE at the current end of the transfer list wouldn't achieve that correct alignment, you can just insert a new XFERLIST_VOID TE first to use up the required padding space. I think this would be a lot nicer to handle in practice (and maybe alleviate some of Simon's concerns here?) because the padding wouldn't be part of the ACPI TE itself anymore, and the code reading that TE wouldn't have to jump through extra hoops to deal with it. (This approach would also require lowering the general TE header alignment to 8, because in order to align the start of the data portion to 16 bytes or higher, the header would have to be 8 bytes in front of that alignment boundary and thus couldn't be 16-byte aligned itself.)

Reply:
> I think if we agree on that mechanism to handle any larger-than-natural alignments, it makes sense to reevaluate whether we even need to stick to the 16 byte TE alignment, or if we want to reduce the general alignment for all TE headers to 8.

Yes, I advocated that too in another comment.

> Instead, we could just mandate that whenever you write such a TE you have to ensure that the data portion starts at the correct alignment. 

I can see how that would work but it seems about the same level of complexity to me and might waste a bit more space with an extra TE header. I think having a reserved/padding field in TEs that have an alignment requirement is fine.

One other thing to check is whether it's important for TL consumers to know the alignment requirement for a particular TE (assuming this is no longer static)? It's not possible to infer the TE alignment requirement just from the amount of padding or the start data address, even though the max alignment for all TEs will now be stored in the TL header. I don't see this being important but if it is then we should add a power-of-2 alignment field to the ACPI TE.

Reply:
Firstly, we should stop talking about ACPI as the only case :-) It is only one of many use cases and there are already others, as I mentioned. Note also that ACPI tables are typically on a 4KB boundary, as is done in both U-Boot and coreboot, so we need to support that. We also have SMBIOS and things like TPM logs which are typically on larger alignment.

The TL is relocated once there is enough memory and before some of the larger things are added. U-Boot currently uses 4KB alignment for this, so that everything appears at sensible addresses.

It seems people are OK with adding the power-of-two alignment in the TL header - good.

I am not sure what "maintain the offset from that alignment when copying" means. Perhaps there were some hyphens missing in Dan's description above. My expectation is that a TL must start at the alignment as specified in the header.  So for example:

0x1240 TL header (align=0x20)
0x1260 TE header (align=1)
0x1270 (0x10 bytes data)
0x1280 (unused)
0x1290 TE header (align=0x20)
0x12a0 (data aligned to 32-byte boundary)

must be copied to at least as large an alignment as it started with, e.g. when adding ACPI:

0x3000 TL header (align=0x1000)
0x3020 TE header (no alignment required)
0x3030 (0x10 bytes data)
0x3040 (unused)
0x3050 TE header (align=0x20)
0x3060 (data aligned to 32-byte boundary)
0x3ff0 TE header (align=0x1000)
0x4000 (data aligned to 4KB boundary)

Note that if TEs are added with larger alignment than given by the TL header, the TL must be updated (and probably relocated) first.

As Julius says, the required data alignment affects the TE-header alignment if we don't use the header size in the way it is currently done. But I believe the TE-header size is the only sensible way to deal with TE alignment (absent another field).

Re the VOID idea, yes we could do that. But then the TE alignment is not specified in any way in that TE, but in the one before, and only by implication / convention. If there were concerns about using the TE-header size to indicate alignment (and anything else might affect where the data actually needs to go) , then surely creating a whole other entry is worse. We should have the information actually present explicitly, if it is needed.

That suggests we might be best to add an alignment field to the TE header also, although I still feel that using the TE-header size for that is also a reasonable option.

Reply:
> Firstly, we should stop talking about ACPI as the only case :-)

OK, fair point.

> I am not sure what "maintain the offset from that alignment when copying" means. Perhaps there were some hyphens missing in Dan's description above. My expectation is that a TL must start at the alignment as specified in the header.

I'm saying the TL does not necessarily need to start at the alignment specified in the header (though it can), but the offset between TL base address and the aligned TL base address must be kept the same if the TL is copied/relocated somewhere else. In this way there's no need to align the TL/TE header base addresses, and it's possible to append new TEs with larger alignments without necessarily relocating the whole TL. The "offset to maintain" would change when adding such TEs. Of course if there's not enough memory allocated for the new TE, then relocation will be needed.

So Simon's example TE might look something like this:

0x1240 TL header (max data alignment = 0x20)
0x1242 TE header (no alignment required)
0x1243 (0x10 bytes data)
0x1253 TE header (data alignment = 0x20)
0x1254 Reserved/padding
0x1260 (data aligned to 32-byte boundary)

The "offset to maintain" here would be 0. When adding ACPI:

0x1240 TL header (max data alignment = 0x1000)
0x1242 TE header (no alignment required)
0x1243 (0x10 bytes data)
0x1253 TE header (data alignment = 0x20)
0x1254 Reserved/padding
0x1260 (data aligned to 32-byte boundary, say 0x20 bytes)
0x1280 TE header (data alignment=0x1000)
0x1281 Reserved/padding
0x2000 (data aligned to 4KB boundary)

Now the "offset to maintain" is 0x240. Alternatively, you could still do the upfront relocation and it would look something like this:

0x3000 TL header (max data alignment = 0x1000)
0x3012 TE header (no alignment required)
0x3013 (0x10 bytes data)
0x3023 TE header (data alignment = 0x20)
0x3024 Reserved/padding
0x3040 (data aligned to 32-byte boundary, let's say 0x20 bytes)
0x3060 TE header (data alignment=0x1000)
0x3061 Reserved/padding
0x4000 (data aligned to 4KB boundary)

Now the "offset to maintain" is 0 but the TL is 0x240 bigger.

> But I believe the TE-header size is the only sensible way to deal with TE alignment (absent another field).

There is another field - e.g. the reserved field in the ACPI TE. That can be renamed if you want. If we overload the TE header size for this, it prevents TE header expansion for new fields.

If Simon strongly objects to putting this reserved/padding/alignment field in the specific TEs and Julius strongly objects to forcing the TE header to be > 8 bytes, then the only alternative I see is to define 2 TE headers now: a "compact" TE header that is guaranteed to always be 8 bytes, and an "extensible" TE header starting at 16 bytes. If we went this way, then we wouldn't need to use a  byte in the "compact" header for the header size; a single bit would do to identify the TE header type, leaving 7 bits for other stuff.

Reply:
> I can see how that would work but it seems about the same level of complexity to me and might waste a bit more space with an extra TE header.

I don't see how it could waste space -- if the general TE header alignment requirement is 8 bytes, and the TE header size is also 8 bytes, then the minimum amount of padding you might need for larger alignments is 8 bytes (that case would be when you need to align to 16 bytes and you're currently at the 8 byte alignment boundary between two 16 byte boundaries). In that case, creating an XFERLIST_VOID header with data_size 0 will do exactly what you need.

I still really think this is the superior solution, because it factors all this alignment complexity out to be a generic thing that you only need to care about when writing (not when reading), and you can treat the same way for all TEs that have a special alignment requirement. You would just have to write one `prepare_for_aligned_te(struct transfer_list *tl, size_t alignment)` function that calculates the appropriate padding size and creates a VOID entry if needed, and then you don't need to add those extra padding fields to each individual TE layout description for each tag affected by this problem. You would just have your `create_acpi_te()` function and all the others call that same prepare function once at the top and that's it. I think that would really encapsulate most of the complexity here a lot nicer than any of the other suggestions, without adding any extra fields to the TE header.

Plus, using VOID entries means that you can later reuse that space to place other smaller TEs if you want to. In that alignment to 4KB example where you're adding thousands of bytes of padding, you could later easily overwrite the VOID entry with another TE and create a new, smaller VOID entry behind it to cover the remaining parts of unused padding, so that the space isn't forever wasted. You can even do that if you don't know how to interpret the actual tag that required the alignment.

I don't think there's any need to store alignment requirements in a TE itself. The only use case we have that cares about the alignment requirement after it is written is relocation, and we can already solve that in a better (more compact) way by just keeping track of the maximum alignment requirement in the TL header, and using Dan's "offset to maintain" approach.

Reply:
I decided to mock up how I think the spec should look according to my latest proposal, hoping that maybe seeing it spelled out like that will help convince you that it's an easier and cleaner approach. I also thought it might be helpful to write down in a sort of algorithm description not-quite-pseudocode how the various operations (e.g. relocation, adding a TE with alignment) should look, because I thought it would make the rest of the spec easier to understand if we clarify how these various fields should be used by example. I think it would also make a good guideline for implementers to help them get all the tricky bits right (e.g. you really gotta pay attention to always keep the checksum correct... you can tell me if I screwed up anywhere in that description when you review it ;) ). I don't know if you want something like that in the specification, if not maybe it could be some sort of appendix instead, but I think it would be a useful thing to have available somewhere. Also, I think it really helps illustrate how easy the whole alignment-by-VOID-TE fits in as just a thin wrapper around the normal "add a TE" algorithm, without requiring any more fields that need to be handled in every other operation. Let me know what you think.

Reply:
> It seems to create all sorts of problems, judging by the length of the discussion above. I'd like to see a clear reason to save these 8 bytes (per TE). Obviously smaller is better in general, but we seem to be creating a rod for our own back, to save maybe 50 or 100 bytes. Julius, could you please have a think about that?

Honestly, Simon, I haven't really understood a single actual practical problem created by this solution or how your proposals solve the same thing in a better way. You keep saying that this was vaguely "bad" somehow but without actually naming the drawback, other than things like having to write one more line of code in some cherry-picked example. Part of my goal with actually writing down step-by-step how each operation would work was so that we could clearly discuss how differences in the data structure design would affect those operations, and compare which things become simpler or more complicated in which model.

So your solution is to add an "alignment" field to the TE header so that each TE stores its own alignment requirement? And then what... that somehow makes everything magically cleaner and simpler and "KISS"? Why? When you specify the alignment this way (instead of fixing the alignment offset to the whole TL header as with Dan's suggestion), then the relocation operation basically means that you have to rebuild the whole TL from scratch one TE at a time and re-align every TE individually according to the requirement stored in its header in the new TL location. In what way is that "simpler" than our "just find a base with the same alignment boundary offset and memcpy() the whole thing in one go" solution? Your proposal is the one that would actually require more complicated and cumbersome (and not to mention slower) code in the end.

> void *bloblist_add(uint tag, int size, int align);
>
> All that is needed is to specify the alignment. We don't need any functions like prepare_for_aligned_te().

You're being facetious here -- obviously my solution could be written to offer the same function prototype as its external API (which is basically the way I wrote it now in the operation description with the latest update -- see the "Adding a new TE with special data alignment requirement" case). The point is that the operation to align a TE is generic and can be easily encapsulated, and that's true for both cases.

> The real question is how much 'reserved' memory do we pass to Linux? The whole TL, or just the parts that have data that Linux needs?

This is a valid question but not one that really matters for this discussion, since no proposal really has any advantage over the other in this matter. Regardless of whether alignment padding is internal to the TE or implemented as separate VOID tags, it's still going to be roughly the same amount of padding and thus roughly the same amount of wasted memory if permanently reserved from the kernel.

Reply:
> > It seems to create all sorts of problems, judging by the length of the discussion above. I'd like to see a clear reason to save these 8 bytes (per TE). Obviously smaller is better in general, but we seem to be creating a rod for our own back, to save maybe 50 or 100 bytes. Julius, could you please have a think about that?
> 
> Honestly, Simon, I haven't really understood a single actual practical problem created by this solution or how your proposals solve the same thing in a better way. You keep saying that this was vaguely "bad" somehow but without actually naming the drawback, other than things like having to write one more line of code in some cherry-picked example. Part of my goal with actually writing down step-by-step how each operation would work was so that we could clearly discuss how differences in the data structure design would affect those operations, and compare which things become simpler or more complicated in which model.

Can we just address the need to save those 50/100 bytes first? That is the hangup I have here.

> 
> So your solution is to add an "alignment" field to the TE header so that each TE stores its own alignment requirement? And then what... that somehow makes everything magically cleaner and simpler and "KISS"? Why? When you specify the alignment this way (instead of fixing the alignment offset to the whole TL header as with Dan's suggestion), then the relocation operation basically means that you have to rebuild the whole TL from scratch one TE at a time and re-align every TE individually according to the requirement stored in its header in the new TL location. In what way is that "simpler" than our "just find a base with the same alignment boundary offset and memcpy() the whole thing in one go" solution? Your proposal is the one that would actually require more complicated and cumbersome (and not to mention slower) code in the end.

I don't understand why we need to rebuild the TL. Dan's suggestion is fine but I don't see the need for it, since we can put the TL on a 16-byte boundary initially and relocate it to a 4KB one later, if needed...?

I am fine to not have an alignment field in the TE header, so long as the header size can handle that bit. But I think you were objecting to that?

> 
> > void *bloblist_add(uint tag, int size, int align);
> > All that is needed is to specify the alignment. We don't need any functions like prepare_for_aligned_te().
> 
> You're being facetious here -- obviously my solution could be written to offer the same function prototype as its external API (which is basically the way I wrote it now in the operation description with the latest update -- see the "Adding a new TE with special data alignment requirement" case). The point is that the operation to align a TE is generic and can be easily encapsulated, and that's true for both cases.

OK I missed that completely. I thought you were still saying that ACPI would have to do special things to work.

> 
> > The real question is how much 'reserved' memory do we pass to Linux? The whole TL, or just the parts that have data that Linux needs?
> 
> This is a valid question but not one that really matters for this discussion, since no proposal really has any advantage over the other in this matter. Regardless of whether alignment padding is internal to the TE or implemented as separate VOID tags, it's still going to be roughly the same amount of padding and thus roughly the same amount of wasted memory if permanently reserved from the kernel.

Agreed. But my point was that wasting this memory is not really a problem and we shouldn't worry about it.

Back to my earlier point, we seem to be hung up on reducing the TE header from 16 bytes to 8. Is there a strong motivation for that?

Reply:
> Can we just address the need to save those 50/100 bytes first? That is the hangup I have here.

Sorry, I don't understand where you're going with this. Clearly, wasting less space on padding is better than wasting more (and in tight SRAM environments, that difference could be significant). If there was a trade-off to be made, if we had to drop significant capabilities in favor of saving that space, then we could try to evaluate whether those savings would be more helpful in practice than what we'd be giving up for them. But I don't see the trade-off here? What is it actually that you're still trying to get into this specification that the proposals we have at this point cannot already provide?

> I don't understand why we need to rebuild the TL. Dan's suggestion is fine but I don't see the need for it, since we can put the TL on a 16-byte boundary initially and relocate it to a 4KB one later, if needed...?

Sorry, I am still not really clear what exact alternative model you're arguing for here, so it's really hard to discuss specific advantages and disadvantages. (Sounds like you don't want alignment fields in the TE headers after all?)

Solving the alignment question means defining two things: how to add a new TE at a specific alignment, and how to relocate a whole TL from one memory area to another without breaking the alignment requirements of all existing TEs. The new section I added (https://github.com/jwerner-chromium/firmware_handoff/blob/main/source/transfer_list.rst#standard-operations) describes how both of those would be solved with my proposal. If you think that's not a good idea, please explain how you think both of those things should be done instead and what's better about that variation.

> I am fine to not have an alignment field in the TE header, so long as the header size can handle that bit. But I think you were objecting to that?

I still think we should add separate VOID TEs to achieve alignments, for all the reasons mentioned (simplifies handling of the TE that needs the alignment, allows adding optional TE header fields in the future if we choose to, allows reusing padding for other TEs if implementations want to). I don't see any real drawbacks with this option.

Reply:
I'll try to respond to the key points in the last few comments...

> [Julius] I don't see how it [using XFERLIST_VOID for padding] could waste space.

Yes, you're right if the TE header is 8 bytes. Even if it were larger, this is probably a negligible concern.

> [Julius] I still really think this is the superior solution, because it factors all this alignment complexity out to be a generic thing ...

Yes, that's a winning argument for me.

> [Julius] Plus, using VOID entries means that you can later reuse that space to place other smaller TEs if you want to.

Nice, though I doubt this would be used in practice as Simon observes.

> [Simon I don't understand why we need to rebuild the TL [if we put an alignment field in the TE header]. Dan's suggestion is fine but I don't see the need for it, since we can put the TL on a 16-byte boundary initially and relocate it to a 4KB one later, if needed...?

That means we'd have to mandate that before adding a TE with an alignment requirement higher than in the TL header, the TL must be first relocated to an address with that new alignment. That matches your intended usage but seems less flexible to me, as well as bloating the TE header for all TEs.

Also, why put the TL on 16-byte boundary initially? Can't this be done just before (relocating and) adding TEs with an alignment requirement?

> [Simon[ At this point I think the key thing to figure out is whether we move to an 8-byte TE header...

You initially mentioned alignment as a reason to keep the TE header at 16-bytes but that no longer seems a good reason. Julius's alignment solution satisfies everyone's with some additional nice-to-have properties.

The TE header size is really a separate thing. It's about whether we want to support the extra complexity of 2 headers sizes (now or in the future) or make everyone pay the overhead now of 16-bytes. I don't think the complexity of using 8-bytes now and expanding to 16-bytes is that bad.

> [Julius] I decided to mock up how I think the spec should look according to my latest proposal

Thanks. I'll check that out soon.

Reply:
It is very hard to predict the future. My point about the TE size is mostly that we are right at the limit (with the reduced tag space to 24 bytes - which actually seems fine to me) so anything people suggest is going to cause it to increase. If there is not much benefit to having it at 8 bytes, then why do it?

Re alignment, I don't see a lot of need to be clever here. We will relocate the structure a few times but always to larger memory space, so it will be fine.

The VOID entries seem OK to me, but only as a way to change to an 8-byte TE header. Otherwise I wouldn't say it is worth it (tiny amount of extra processing of a useless header). Let's be clear that this is not explicit alignment. It is just as implicit as the 'header size' thing is...in fact it is even more implicit. I don't see that being a problem, though.

My vote is to stick with 16 byte TE headers since it derisks things. But I don't see a particular need for space in the TE header if we are not storing the alignment there. Can't predict the future and I'd hate to have to increase it back to 16 bytes later, which I why I keep asking why it matters so much?! Not that many months ago the EFI people were wanting a 16-byte UUID field in each TE header :-)

Reply:
Hi Simon.

I don't think it's helpful to conflate the alignment solution and the need for a 8/16 byte TE header. As Julius says, I don't think you (Simon) have described a viable alternative alignment solution. We've already (I hope) discounted overloading the TE hdr_size to align TE data, so I guess you're advocating at least 1 more field in the TE header? This could, for example, be a "data offset" field indicating the offset from the end of the header to the start of the data, but then the alignment is still "implicit". I think managing such a field would be equivalent in complexity as using VOID TEs, but wouldn't have the other nice-to-have properties Julius mentioned. So why add it? It might actually restrict us from extending the alignment use-cases later (see next paragraph).

Alternatively, the extra TE header field could be an explicit power-of-2 data alignment (from the end of the TE header). In theory this would allow TL relocation to arbitrary addresses but that may require complex shuffling around of TEs within the TL, and initializing the padding or removing it from the checksum. I don't think anyone's advocating this use-case so I don't see how adding such a field now would help. We'd still have to define similar rules as Julius has around safe relocation and adding TEs (to avoid that shuffling). Better to add this as and when it's needed.

Separately, there's my idea of "maintaining the TL offset from the max alignment", which gives some extra nice-to-have properties. I guess this is the bit where you think we don't need to be clever? Perhaps, but we would still need to define rules around safe TL relocation and to ban adding TEs with a larger alignment until relocation to that new alignment. If we go that way, then I no longer see a strong need for the alignment field in the TL header.

For any solution, we need to change the rules around 16-byte alignment. I don't see why we need to start with 16-byte as you earlier suggested.

Finally, there's the 8/16 byte TE header thing. I'm pretty neutral about this but it seems to be the stalling point for progress. I don't mind if we start with 8 bytes and expand later, go with 16 bytes or define both now. Perhaps Julius and you can agree privately about this?

Reply:
The alignment suggestion I mentioned above was to have an alignment field in the TE which explicitly indicates the alignment, e.g. as a log2 value. The normal value of 0 would indicate there is no particular alignment. That is the second option you have above. As you say I am not advocating being able to move the TL except by using a simple memcpy() and fixup of TL header.

Re paragraph 3, it could be useful as I said, but I feel that it is simpler to have  the TL start on an aligned boundary (e.g. 4KB) since you are relocating it anyway. One reason for an alignment field in the TL header is that when relocating, a subsequent phase can honour (at least) that alignment. Right now that would have to be implicit, which is a little risky. I feel that this alignment thing is a hole in what I original proposed (some moons ago now), since I had assumed it could be used in that way, but had not made it explicit.

Re the 8/16 byte TE header, I'll connect with Julius on this and see if we can figure it out. So long as we don't actually need 16 bytes, I'm OK to move to 8 bytes. Changing later would reduce credibility for this spec. But how can we know? It just seems like an unnecessary risk to me, given that we have not come up with any real benefit of saving those 50-or-so bytes in early firmware.

Reply:
> One reason for an alignment field in the TL header is that when relocating, a subsequent phase can honour (at least) that alignment.

Right, but have you considered how that relocation algorithm would look in practice? You would have to manually relocate each TE one by one, inspect their alignment fields, find a new aligned location for them in the relocated TL... you're basically rebuilding the whole TL from scratch one entry at a time. Where's the advantage in that? Why are you in favor of this solution? Dan's solution allows us to perform the same relocation in a single memcpy() for the whole TL, _and_ it takes up less space per TE to boot. I see only advantages on one side and only disadvantages on the other here, so I'm confused why it is still controversial.

> So long as we don't actually need 16 bytes, I'm OK to move to 8 bytes. Changing later would reduce credibility for this spec. But how can we know? It just seems like an unnecessary risk to me

I am still very confused about your concern here. How would it "reduce credibility" if we updated the spec in a way that it was designed from the start to be updated? What difference would it make for the case of adding a new field whether the header was already 16 byte to start with or not? Even if we right now said that the header should have an additional 8 bytes of "reserved" fields at the end, it would still require a later update to the spec to actually redefine and add meaning to those fields. So I'm very confused why you think that a spec update that declares "the reserved field at offset 0x8 in the TE header now means this" was somehow "more credible" than a spec update that declares "the TE header is now 16 bytes long with a new field added at offset 0x8 that means this".

I'm happy to have that meeting if you want, but I can tell you right now that I really don't understand this line of thinking. Size _does_ matter for all data structures, and particularly ones meant for use in pre-RAM environments. There's no point about arguing whether it will "only" be XX-or-so bytes as long as the proposal on the other side is just "waste more bytes for no reason with no tangible advantage". With all other things being equal, smaller is always better.

Reply:
1. I think we covered that before. You don't need to rebuild the thing TE by TE. Where are you getting that from?
2. Changing the TE header size seems like something we should avoid if we can, at least until something dramatic changes. The suggestion here is not 8 bytes of reserved space, it is 4, since the header size, tag and data size each consume 4 bytes. We can press a reserved word into server fairly easily, but updating the header size seems more dramatic to me.  It just seems risky to leave absolutely no room at this early stage of the spec.

Thanks for addressing the size question. This is simply a difference of perspectives. From your side, I believe you want the smallest possible data structure, since you expect to add a 5-10 TEs to it early in firmware. From my side I want something that scales from early firmware to late firmware, where flexibility is more important. That said, I don't have specific things that are lacking in what we have (or I would have mentioned them), just as ( I believe) you don't have specific reasons why saving those 50-100 bytes is so important. Bear in mind that the code for this handling this data structure is around 1KB on Thumb2.

As to credibility, it doesn't look to change the header size just after people start using the spec. I believe that is self-evident. If it lasts a few years, that is fine. As I said, so long as we don't actually need 16 bytes, I am OK to move to 8. But that is the question.

Reply:
> I think we covered that before. You don't need to rebuild the thing TE by TE. Where are you getting that from?

Sorry, maybe I misread the part I responded to. I thought you were still discussing "alignment field in TE header" (because if you have that then relocating means reading all these fields in each TE), but I see now that the part I quoted actually said "alignment field in the TL header". So are we actually on the same page for alignment now, then? Because the solution that Dan and I are proposing is the one with a field in the TL header.

> Changing the TE header size seems like something we should avoid if we can, at least until something dramatic changes. [...] We can press a reserved word into server fairly easily, but updating the header size seems more dramatic to me. It just seems risky to leave absolutely no room at this early stage of the spec.

I still really don't follow here. You just keep using vague terms like "dramatic", "risky"... what does that actually mean? What exact risk are you worried about? We have a spec that allows for variable header sizes so what's risky about using that? Why would we even have provisioned for that option if you don't think it's something we should use?

> As to credibility, it doesn't look to change the header size just after people start using the spec. I believe that is self-evident.

Again, I'm completely not following here. If you're trying to argue that it doesn't look good to update the spec at all immediately after people start using it, then okay, I could see that argument. But where's the difference between allocating meaning to a reserved word and expanding a data structure? It's the same thing! Expanding a variable-sized structure is just starting to use the infinite amount of reserved words beyond its end.

> As I said, so long as we don't actually need 16 bytes, I am OK to move to 8. But that is the question.

FWIW I don't expect that we'll actually expand this header anytime soon (or ever), I'm just explaining that it wouldn't be a problem if we had to because you seem to be concerned that some currently unseen need comes up. I think the fields we have defined right now are really everything a generic tagged list structure like this should ever need, and I really can't think of anything else that would be generically useful and necessary to every use case. The only things I expect might get added at some point are edge case things specific to the needs of a small subset of users, and in those scenarios I would then probably advocate to make those fields optional (with the header size not actually increasing for TEs that don't use them) so that other users don't get impacted.

Reply:
Julius and I met to discuss this. We agreed this

1. Change to the 8-byte TE header as Julius proposes. It solves the current need, is more efficient and provides still provides for alignment. Neither Julius nor I can think of a problem with this header at present.
2. While projects are implementing this spec over the coming months, if anything comes up that makes this header untenable we will review it and potentially change if needed. In that case it will be early enough that we won't require the 'old' header to be supported for backwards compatibility.

Julius is going to split out the major (agreed) part of this PR so we can get it submitted.

Julius please correct the above if needed.

Reply:
Julius, any update on this?

Reply:
Apologies for the radio silence, I had (and unfortunately still have) to deal with an unexpected high-priority issue the last two weeks which didn't really leave me any time to look at this again.

I've cut off the last two patches from this PR now, if I understood correctly we have agreement on the remaining 4 so I hope we can merge them. After that, I can upload the other two again to continue the discussions there.

Reply:
I've uploaded the two patches that were still open here as https://github.com/FirmwareHandoff/firmware_handoff/pull/6 and https://github.com/FirmwareHandoff/firmware_handoff/pull/7 now. Please continue review there.